sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ddl.scala

```diff
@@ -190,21 +194,19 @@ private[sql] class DDLParser(
     }
 }
 
-private[sql] object ResolvedDataSource {
-
-  private val builtinSources = Map(
-    "jdbc" -> "org.apache.spark.sql.jdbc.DefaultSource",
-    "json" -> "org.apache.spark.sql.json.DefaultSource",
-    "parquet" -> "org.apache.spark.sql.parquet.DefaultSource",
-    "orc" -> "org.apache.spark.sql.hive.orc.DefaultSource"
-  )
+private[sql] object ResolvedDataSource extends Logging {
 
   /** Given a provider name, look up the data source class definition. */
   def lookupDataSource(provider: String): Class[_] = {
     val loader = Utils.getContextOrSparkClassLoader
-
-    if (builtinSources.contains(provider)) {
-      return loader.loadClass(builtinSources(provider))
+    val sl = ServiceLoader.load(classOf[DataSourceProvider], loader)
+
+    sl.iterator().filter(_.format() == provider).toList match {
+      case Nil => logInfo(s"provider: $provider is not registered in the service loader")
+      case head :: Nil => return head.getClass
+      case sources => sys.error(s"Multiple sources found for $provider, " +
```

Comment:
Is printing an error enough? If it is this should most probably be `logError`. But it feels like this should be an exception.